geode-gfsh/src/main/java/org/apache/geode/management/internal/cli/functions/ReplicateRegionFunction.java

```diff
@@ -0,0 +1,386 @@
+    final boolean isCancel = (Boolean) args[2];
+    long maxRate = (Long) args[3];
+    int batchSize = (Integer) args[4];
+
+    final InternalCache cache = (InternalCache) context.getCache();
+    final Region region = cache.getRegion(regionName);
+    if (isCancel) {
+      return cancelReplicateRegion(context, region, senderId);
+    }
+
+    GatewaySender sender = cache.getGatewaySender(senderId);
+    if (!sender.isRunning()) {
+      return new CliFunctionResult(context.getMemberName(), CliFunctionResult.StatusState.ERROR,
+          "Sender not running");
+    }
+
+    if (!sender.isParallel() && !((InternalGatewaySender) sender).isPrimary()) {
+      return new CliFunctionResult(context.getMemberName(), CliFunctionResult.StatusState.OK,
+          "Sender is serial and not primary. 0 entries replicated.");
+    }
+
+    try {
+      return executeReplicateFunctionInNewThread(context, region, sender, maxRate, batchSize);
+    } catch (InterruptedException e) {
+      return new CliFunctionResult(context.getMemberName(), CliFunctionResult.StatusState.ERROR,
+          "Execution canceled.");
+    } catch (ExecutionException e) {
+      return new CliFunctionResult(context.getMemberName(), CliFunctionResult.StatusState.ERROR,
+          "Execution failed: " + e.getMessage());
+    }
+  }
+
+  private CliFunctionResult executeReplicateFunctionInNewThread(FunctionContext<String[]> context,
+      Region region, GatewaySender sender, long maxRate, int batchSize)
+      throws InterruptedException, ExecutionException {
+    Callable<CliFunctionResult> callable =
+        new ReplicateRegionCallable(context, region, sender, maxRate, batchSize);
+    FutureTask<CliFunctionResult> futureTask = new FutureTask<>(callable);
+    ExecutorService executor = LoggingExecutors
+        .newSingleThreadExecutor(getReplicateRegionFunctionThreadName(region.getName(),
+            sender.getId()), true);
+    executor.execute(futureTask);
+    return futureTask.get();
+  }
+
+  class ReplicateRegionCallable implements Callable<CliFunctionResult> {
+    private final FunctionContext<String[]> context;
+    private final Region region;
+    private final GatewaySender sender;
+    private final long maxRate;
+    private final int batchSize;
+
+    public ReplicateRegionCallable(final FunctionContext<String[]> context, final Region region,
+        final GatewaySender sender, final long maxRate,
+        final int batchSize) {
+      this.context = context;
+      this.region = region;
+      this.sender = sender;
+      this.maxRate = maxRate;
+      this.batchSize = batchSize;
+    }
+
+    @Override
+    public CliFunctionResult call() throws Exception {
+      return replicateRegion(context, region, sender, maxRate, batchSize);
+    }
+  }
+
+  private CliFunctionResult replicateRegion(FunctionContext<String[]> context, Region region,
+      GatewaySender sender, long maxRate, int batchSize) {
+    final Set<?> entries = getEntries(region, sender);
```

Comment:
Maybe you could get the entries right before you need them, before initializing the iterator at line 221. If the `senderPool` or the `connection` are `null`, you will have got the entries for nothing.

Reply:
Good point. Thanks!